### PR TITLE
Rfc/git 1630 signals

### PR DESCRIPTION
**Pull request by ahopkins**
Take over from #1770 re: #1630

### CODE REVIEW

**Comment by ashleysommer** [comment]
@ahopkins 
Thanks for taking this over. I'm very interested in this Signals work, because it is integral for the future of Sanic-Plugins-Framework.
I'll read over the proposed implementation and see if I can offer feedback.

File: sanic/app.py [hunk 1]
```diff
@@ -146,7 +152,7 @@ def register_listener(self, listener, event):
         :param event: when to register listener i.e. 'before_server_start'
         :return: listener
         """
-
+        subscribe(event_name=event, callback=listener, signals=self.signals)
```
Comment (ashleysommer) [inline sanic/app.py:155]: If self.listener() decorator calls `subscribe()` (on line 141 above) then why does this helper function also need to call `subscribe()`? Wouldn't this result in two subscribes with the same listener and signals?

File: sanic/app.py [hunk 2]
```diff
@@ -219,9 +225,7 @@ def response(handler):
         return response
 
     # Shorthand method decorators
-    def get(
-        self, uri, host=None, strict_slashes=None, version=None, name=None
-    ):
+    def get(self, uri, host=None, strict_slashes=None, version=None, name=None):
```
Comment (ashleysommer) [inline sanic/app.py:228]: Looks like quite a lot of the changes in this file are formatting changes, was that just a result of merging in latest master? Or should these formatting changes be filtered out?

File: sanic/helpers.py [hunk 3]
```diff
@@ -154,3 +188,44 @@ def import_string(module_name, package=None):
     if ismodule(obj):
         return obj
     return obj()
+
+
+def _extract_signal_namespace(event_name, signals):
+    global _CLASSIC_EVENT_ALIAS
```
Comment (ashleysommer) [inline sanic/helpers.py:194]: can this be done without using a global variable?

File: sanic/helpers.py [hunk 4]
```diff
@@ -3,6 +3,9 @@
 from importlib import import_module
 from inspect import ismodule
 
+from sanic.signals import Namespace
```
Comment (ashleysommer) [inline sanic/helpers.py:6]: Looks like `Namespace` is only imported for Type checking, so has no purpose at runtime.

This import can changed to:
```
if typing.TYPE_CHECKING:
    from sanic.signals import Namespace
```
Then in type declarations, change `Namespace` to `'Namespace'` (string form)

File: sanic/app.py [hunk 5]
```diff
@@ -92,6 +87,18 @@ def __init__(
         self.go_fast = self.run
         self.test_mode = False
 
+        # Signal Handlers and Registry
+        self._signals = {
+            "server": Namespace(namespace="server", owner=self, classic_event=True,),
+            "request": Namespace(namespace="request", owner=self),
+            "response": Namespace(namespace="response", owner=self),
+            "middleware": Namespace(namespace="middleware", owner=self),
+        }
```
Comment (ashleysommer) [inline sanic/app.py:96]: Looks like 4 Namespaces of signals are defined, but signal emitters are only implemented for `server` signals. Are the others planned to be implemented?

**Comment by ahopkins** [comment]
I think this is a really great potential win. I've been playing with the branch, and have some ideas on how to refine the implementation. I'll post on the RFC thread some changes I'm thinking about. One is to scrap the additional dependency and just roll our own signaling solution.

**Comment by ahopkins** [comment]
Coming 2021

**Pull request closed without merging**